ICSharpCode.Decompiler/CSharp/ProjectDecompiler/WholeProjectDecompiler.cs

```diff
@@ -587,23 +598,47 @@ static string CleanUpFileName(string text, bool separateAtDots)
 			if (pos > 0)
 				text = text.Substring(0, pos);
 			text = text.Trim();
-			// Whitelist allowed characters, replace everything else:
-			StringBuilder b = new StringBuilder(text.Length);
+			string extension = null;
 			int currentSegmentLength = 0;
 			var (supportsLongPaths, maxPathLength, maxSegmentLength) = longPathSupport.Value;
+			if (treatAsFileName)
+			{
+				// Check if input is a file name, i.e., has a valid extension
+				// If yes, preserve extension and append it at the end.
+				// But only, if the extension length does not exceed maxSegmentLength,
+				// if that's the case we just give up and treat the extension no different
+				// from the file name.
+				int lastDot = text.LastIndexOf('.');
+				if (lastDot >= 0 && text.Length - lastDot < maxSegmentLength)
+				{
+					extension = text.Substring(lastDot);
+					text = text.Remove(lastDot);
+					foreach (var c in extension)
+					{
+						if (!(char.IsLetterOrDigit(c) || c == '-' || c == '_' || c == '.'))
+						{
+							// extension contains an invalid character, therefore cannot be a valid extension.
+							extension = null;
```

Comment:
We just drop the whole extension?
Don't we need to revert to the original `text`?

Reply:
ah, yes... thanks!